### PR TITLE
JDBC Tasks of Marian Zlatev Tasks 4

### PR DESCRIPTION
**Pull request by NewestUser**
Created different repositories for accessing and manipulating data from users, contacts and addresses DB tables.

### CODE REVIEW

File: tests/com/clouway/jdbc/repositories/ContactRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  public void bindingAUserToAddressAndLookingUpResult() throws Exception {
+    addressRepository.store(new Address(1, "Konstantin panica"));
+    userRepository.store(new User(1, "Ivailo"));
+    contactRepository.store(new Contact("Ivailo", "Konstantin panica"));
+
+    List<Contact> allContacts = contactRepository.lookupAll();
+    assertThat(allContacts.get(0).getName(), is(equalTo("Ivailo")));
+  }
+
+  private void clearTablesContent() {
```
Comment (mgenov) [inline tests/com/clouway/jdbc/repositories/ContactRepositoryTest.java]: seems too coupled with the implementation. You probably could drop existing database and re-create it from the schema. That way you will for the test to use the production ready schema and duplication will be moved from test to production code.

File: tests/com/clouway/jdbc/repositories/UserRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+public class UserRepositoryTest {
+
+  private UserRepository userRepository;
+  private Connection conn;
+
+  @Before
+  public void setUp() throws Exception {
+    conn = DriverManager.getConnection("jdbc:mysql://localhost:3306/examples", "root", "password");
+    userRepository = new UserRepository(conn);
+    userRepository.rawUpdate("SET FOREIGN_KEY_CHECKS = 0");
```
Comment (mgenov) [inline tests/com/clouway/jdbc/repositories/UserRepositoryTest.java]: . 

Note that rawUpdate method does not belongs to the UserRepository cause User is concrete where rawUpdate is common.

File: src/com/clouway/jdbc/repositories/ContactRepository.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  public ContactRepository(Connection conn) {
+    this.connection = conn;
+  }
+
+  @Override
+  public void store(Contact value) {
+    try {
+      ResultSet mappingResult = mapUserAndAddress(value.getName(), value.getAddress());
+      mappingResult.next();
+      int usrId = mappingResult.getInt(1);
```
Comment (mgenov) [inline src/com/clouway/jdbc/repositories/ContactRepository.java]: userId 
addressId

File: tests/com/clouway/jdbc/repositories/UserManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import static org.junit.Assert.*;
+
+public class UserManagerTest {
+
+  private UserManager manager;
+  private ContactRepository contactRep;
+
+  @Before
+  public void setUp() throws Exception {
+    Connection conn = DriverManager.getConnection("jdbc:mysql://localhost:3306/examples", "root", "password");
```
Comment (mgenov) [inline tests/com/clouway/jdbc/repositories/UserManagerTest.java]: single rule could be used for accessing database connection between tests.

File: src/com/clouway/jdbc/repositories/ContactRepository.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        Contact item = fetcher.fetchRow(resultSet);
+        results.add(item);
+      }
+    } catch (SQLException e) {
+      e.printStackTrace();
+    }
+    return results;
+  }
+
+  private RowFetcher<Contact> getFetcher() {
```
Comment (mgenov) [inline src/com/clouway/jdbc/repositories/ContactRepository.java]: What will happen when multiple threads are using the same instance of the Repository ?

Comment (NewestUser) [inline src/com/clouway/jdbc/repositories/ContactRepository.java]: I think that the fetcher might return inappropriate data for the query currently being executed. But maybe that will only happen if a WHERE clause is added. If one thread executes a lookupAll() with one query and another thread executes a lookupAll() with another query the first thread maybe will get some results from the second query.
In the current situation we are looking up all the contacts.
Maybe in the worst case duplication of entries might occur (or so I think).

File: src/com/clouway/jdbc/addressbook/repositories/Repository.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * @author Marian Zlatev (mzlatev91@gmail.com)
+ */
+public interface Repository<T> {
+
+  void store(T value);
+
+  List<T> lookupAll();
+
+  void update(String query);
```
Comment (mgenov) [inline src/com/clouway/jdbc/addressbook/repositories/Repository.java]: doesn't belongs here, cause it exposes SQL related functionality to the caller.

Comment (NewestUser) [inline src/com/clouway/jdbc/addressbook/repositories/Repository.java]: Accidentally closed all my pull requests. :(

**Pull request merged**